zrml/prediction-markets/src/migrations.rs

```diff
@@ -63,84 +72,66 @@ pub struct OldMarket<AI, BN, M> {
         if market_commons_version != MARKET_COMMONS_REQUIRED_STORAGE_VERSION {
             log::info!(
-                "RecordBonds: market-commons version is {:?}, but {:?} is required",
+                "AddOutsiderBond: market-commons version is {:?}, but {:?} is required",
                 market_commons_version,
                 MARKET_COMMONS_REQUIRED_STORAGE_VERSION,
             );
             return total_weight;
         }
-        log::info!("RecordBonds: Starting...");
+        log::info!("AddOutsiderBond: Starting...");
 
-        let new_markets = storage_iter::<OldMarketOf<T>>(MARKET_COMMONS, MARKETS)
-            .map(|(key, old_market)| {
-                total_weight = total_weight.saturating_add(T::DbWeight::get().reads(1));
-                let creation = Some(match old_market.creation {
-                    MarketCreation::Advised => Bond {
-                        who: old_market.creator.clone(),
-                        value: T::AdvisoryBond::get(),
-                        is_settled: old_market.status != MarketStatus::Proposed,
-                    },
-                    MarketCreation::Permissionless => Bond {
-                        who: old_market.creator.clone(),
-                        value: T::ValidityBond::get(),
-                        is_settled: matches!(
-                            old_market.status,
-                            MarketStatus::Resolved | MarketStatus::InsufficientSubsidy,
-                        ),
-                    },
-                });
-                let oracle = Some(Bond {
-                    who: old_market.creator.clone(),
-                    value: T::OracleBond::get(),
-                    is_settled: matches!(
-                        old_market.status,
-                        MarketStatus::Resolved | MarketStatus::InsufficientSubsidy,
-                    ),
-                });
-                let new_market = Market {
-                    creator: old_market.creator,
-                    creation: old_market.creation,
-                    creator_fee: old_market.creator_fee,
-                    oracle: old_market.oracle,
-                    metadata: old_market.metadata,
-                    market_type: old_market.market_type,
-                    period: old_market.period,
-                    scoring_rule: old_market.scoring_rule,
-                    status: old_market.status,
-                    report: old_market.report,
-                    resolved_outcome: old_market.resolved_outcome,
-                    dispute_mechanism: old_market.dispute_mechanism,
-                    deadlines: old_market.deadlines,
-                    bonds: MarketBonds { creation, oracle },
-                };
-                (key, new_market)
-            })
-            .collect::<Vec<_>>();
+        let mut translated = 0u64;
+        zrml_market_commons::Markets::<T>::translate::<OldMarketOf<T>, _>(|_key, old_market| {
```

Comment:
I am not sure, if I should rather use [translate_values](https://github.com/paritytech/substrate/blob/8d17a16a5653974ec17ed617306606e9dbb6e4b4/frame/support/src/storage/mod.rs#L1232-L1264). I used `translate`, because of [this example](https://github.com/paritytech/substrate/blob/7ba1ef526fbcc224d3bd446c83ee66a83cacec49/frame/assets/src/migration.rs#L70-L79).

I read, that `translate` has the downside, that it silently ignores wrong decodings. Means, when the `OldMarketOf<T>` does not match the storage structure currently present before the runtime upgrade, then we wouldn't notice and it would not change the values. This means a check with `try-runtime` `pre-upgrade` and `post-upgrade` is mandatory to check that the storage was correctly translated! In example one could check that all markets are still present ([count all markets](https://github.com/paritytech/substrate/blob/7ba1ef526fbcc224d3bd446c83ee66a83cacec49/frame/assets/src/migration.rs#L102-L105)) and if [all values are as expected](https://github.com/paritytech/substrate/blob/7ba1ef526fbcc224d3bd446c83ee66a83cacec49/frame/assets/src/migration.rs#L116-L118) after the runtime upgrade.

Reply:
```rust
        let old_markets = storage_key_iter::<MarketIdOf<T>, OldMarketOf<T>, Blake2_128Concat>(
            MARKET_COMMONS,
            MARKETS,
        )
        .collect::<BTreeMap<_, _>>();
        Self::set_temp_storage(old_markets, "old_markets");
```

In combination with the above (`pre_upgrade`), this is dangerous. If `OldMarketOf<T>` is the wrong structure for the old markets, then the `BTreeMap` would be empty and counter is zero.

Assume `translate` fails for each element, because of the wrong decoding struct. From [here](https://paritytech.github.io/substrate/master/frame_support/storage/types/struct.StorageMap.html#method.translate) `NOTE: If a value fail to decode because storage is corrupted then it is skipped.`, then we have the same storage structure as before the upgrade, but the new decoding fails then (no items in Markets). In this particular case this means, that Markets is empty with the new decoding structure.

To mitigate this is to count the items inside `Markets` and assert against `Markets` counter after the migration.

Reply:
The mistake would be noticeable now with the log error from this commit https://github.com/zeitgeistpm/zeitgeist/pull/903/commits/89d75f8c51ac06b737e1ac9b0ca18197475043eb